api/gateway/server.go

```diff
@@ -2,66 +2,68 @@ package gateway
 
 import (
 	"context"
-	"fmt"
 	"net"
 	"net/http"
+	"sync/atomic"
 	"time"
 
 	"github.com/gorilla/mux"
 )
 
-// Server represents an RPC server on the Node.
-// TODO @renaynay: eventually, gateway server should be able to be toggled on and off.
+// Server represents a gateway server on the Node.
 type Server struct {
-	cfg Config
-
 	srv      *http.Server
 	srvMux   *mux.Router // http request multiplexer
 	listener net.Listener
+	started  atomic.Bool
 }
 
-// NewServer returns a new RPC Server.
-func NewServer(cfg Config) *Server {
+// NewServer returns a new gateway Server.
+func NewServer(address string, port string) *Server {
 	srvMux := mux.NewRouter()
 	srvMux.Use(setContentType)
 
 	server := &Server{
-		cfg:    cfg,
 		srvMux: srvMux,
 	}
 	server.srv = &http.Server{
+		Addr:    address + ":" + port,
 		Handler: server,
 		// the amount of time allowed to read request headers. set to the default 2 seconds
 		ReadHeaderTimeout: 2 * time.Second,
 	}
 	return server
 }
 
-// Start starts the RPC Server, listening on the given address.
+// Start starts the gateway Server, listening on the given address.
 func (s *Server) Start(context.Context) error {
-	listenAddr := fmt.Sprintf("%s:%s", s.cfg.Address, s.cfg.Port)
-	listener, err := net.Listen("tcp", listenAddr)
-	if err != nil {
-		return err
+	if s.started.CompareAndSwap(false, true) {
+		listener, err := net.Listen("tcp", s.srv.Addr)
+		if err != nil {
+			return err
+		}
+		s.listener = listener
+		log.Infow("gateway: server started", "listening on", listener.Addr().String())
+		//nolint:errcheck
+		go s.srv.Serve(listener)
+		return nil
 	}
-	s.listener = listener
-	log.Infow("RPC server started", "listening on", listener.Addr().String())
-	//nolint:errcheck
-	go s.srv.Serve(listener)
+	log.Warn("gateway: cannot start server: already started")
 	return nil
 }
 
-// Stop stops the RPC Server.
-func (s *Server) Stop(context.Context) error {
-	// if server already stopped, return
-	if s.listener == nil {
+// Stop stops the gateway Server.
+func (s *Server) Stop(ctx context.Context) error {
+	if s.started.CompareAndSwap(true, false) {
+		err := s.srv.Shutdown(ctx)
+		if err != nil {
+			return err
+		}
+		log.Info("gateway: server stopped")
```

Comment:
Lee Ghandi

Reply:
what is Lee Ghandi?

Reply:
https://www.youtube.com/watch?v=-15VC4Yxzys

Reply:
@liamsi wen HR department?

Reply:
dies of ligma